File: apps/web/tools/essay_content.py
"""Read and write essay data plus its byte-exact public Markdown twin.

Frontmatter values are JSON-compatible YAML scalars. Keeping each value on one
line lets the authoring tools use Python's standard library alone.
"""

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
CONTENT = ROOT / "content" / "essays"
PUBLIC = ROOT / "public"
SITE = "https://www.mauricekleine.com"

OPTIONAL_FIELDS = ("inlineLinks",)

FIELDS = (
    "slug", "title", "seoTitle", "summary", "date", "dateDisplay",
    "cover", "coverOriginal", "coverSrcset", "coverSizes", "coverWidth",
    "coverHeight", "coverAlt", "x", "linkedin", "images",
)


def write_content(slug, metadata, markdown):
    """Write data frontmatter followed by unchanged Markdown bytes."""
    if not set(FIELDS).issubset(metadata) or not set(metadata).issubset(set(FIELDS) | set(OPTIONAL_FIELDS)) or metadata["slug"] != slug:
        raise ValueError(f"invalid essay metadata fields: {slug}")
    lines = ["---\n"]
    for key in (*FIELDS, *(key for key in OPTIONAL_FIELDS if key in metadata)):
        lines.append(f"{key}: {json.dumps(metadata[key], ensure_ascii=False)}\n")
    lines.append("---\n")
    path = CONTENT / f"{slug}.md"
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes("".join(lines).encode() + markdown)
    return path


def read_content(path):
    data = path.read_bytes()
    if not data.startswith(b"---\n"):
        raise ValueError(f"missing frontmatter: {path}")
    closing = data.find(b"\n---\n", 4)
    if closing < 0:
        raise ValueError(f"missing closing delimiter: {path}")
    metadata = {}
    for line in data[4:closing].decode().splitlines():
        key, value = line.split(": ", 1)
        metadata[key] = json.loads(value)
    if not set(FIELDS).issubset(metadata) or not set(metadata).issubset(set(FIELDS) | set(OPTIONAL_FIELDS)):
        raise ValueError(f"invalid essay metadata fields: {path}")
    return metadata, data[closing + len(b"\n---\n"):]


def essays():
    """Newest essay first, with slug breaking same-date ties."""
    return sorted(
        (read_content(path)[0] for path in CONTENT.glob("*.md")),
        key=lambda essay: (essay["date"], essay["slug"]),
        reverse=True,
    )

File: apps/web/src/essay-content.ts
import YAML from 'yaml'

type Essay = {
  slug: string
  title: string
  seoTitle: string
  summary: string
  date: string
  order: number
  dateDisplay: string
  cover: string
  coverSrcset: string
  coverSizes: string
  coverAlt: string
  coverWidth: number
  coverHeight: number
  headHtml: string
  beforeArticleHtml: string
  articleHtml: string
  afterArticleHtml: string
  markdown: string
}

const sources = import.meta.glob('../content/essays/*.md', { eager: true, query: '?raw', import: 'default' }) as Record<string, string>

function parseEssay(source: string): Essay {
  const match = /^---\r?\n([\s\S]*?)\r?\n---\r?\n/.exec(source)
  if (!match) throw new Error('essay has no frontmatter')
  const metadata = YAML.parse(match[1]) as Omit<Essay, 'markdown'>
  return { ...metadata, markdown: source.slice(match[0].length) }
}

export const essays = Object.values(sources).map(parseEssay).sort((a, b) => a.order - b.order)
export const essaysBySlug = Object.fromEntries(essays.map((essay) => [essay.slug, essay]))

export function essayBody(essay: Essay) {
  return `${essay.beforeArticleHtml}<article class="essay">\n${essay.articleHtml}\n      </article>${essay.afterArticleHtml}`
}
